### ISSUE TITLE
Preexisting record types

### ISSUE THREAD

**Issue by skabbes** [dropbox/djinni#52]
Djinni really only needs to know how to convert record types between objc, java and c++.  What if we extended the include format to take plugins like this?

@include "date.json"

```coffeescript
djinni_type: "custom_date"
objc_type: "NSDate"
java_type: "java.util.Date"
cpp_include_path: "<chrono>"
cpp_type: "std::chrono::system_clock::time_point"
jni_type: "jobject"
```

```cpp
// custom_date.hpp
namespace djinni {
  jobject custom_date_to_java(time_point);
  NSDate * custom_date_to_objc(time_point);
  time_point custom_date_to_cpp(jobject);
  time_point custom_date_to_cpp(NSDate *);
}
```

**Comment by mknejp** [comment]
This would certainly help making stuff more modular. We are currently struggling with how to setup Djinni in na environment with libraries. Say library A lives in its own repository and has Djinni-generated classes, and library B uses library A and has its own djinni-generated classes that need to interact with A in their interface, then there is no *easy* way to make this work without exposing A's djinni files and hope for no linker errors or ODR violations.

So bonus points if Djinni was able to generate these JSON files for inclusion of generated interfaces in other projects.

**Comment by pwais** [comment]
djinni currently deep-copies all record types.  It would be nice if this proposed feature didn't *necessarily* require deep copies.  E.g. a user could leverage it to effectively share state across the language boundary (much like interfaces do currently).  Shouldn't be challenging to support if the user-provided type mapping function can have arbitrary behavior.

+1 as using this proposed feature to solve #54 !

**Comment by mknejp** [comment]
So, I was on a coding spree and this is now a thing. The thing can be found here: https://github.com/mknejp/djinni/tree/feature/extern-types and is based on the refactoring done in #54. It was actually pretty easy to add once all the necessary machinery was in place.

This contains two main features:
* The ability for Djinni to dump information about the types its generating into a config file and
* the ability to read such files back in and use the stored information for code generation.

I decided to go with YAML instead of JSON for these main reasons:
* YAML can store comments.
* YAML can store multiple documents in a single file, so you can have a single `mylib.yaml` file containing all the types in one place.
* JSON is a subset of YAML so it can still be read if really needed.
* *Much* more human-friendly format

YAML generation is controlled by the following options:
* `--yaml-out`: The output folder for YAML files (Generator disabled if unspecified).
* `--yaml-out-file`: If specified all types are merged into a single YAML file instead of generating one file per type (relative to `--yaml-out`).
* `--yaml-prefix`: The prefix to add to type names stored in YAML files (default: "").

To use external types reference the YAML file using the directive `@extern "mylib.yaml"` (instead of `@import "mylib.yaml"`) and use the names as you would otherwise (don't forget the prefix).

This is an example YAML file generated by Djinni:
```yml
# AUTOGENERATED FILE - DO NOT MODIFY!
# This file generated by Djinni from record1.djinni, interface1.djinni, enum1.djinni
---
name: mylib_record1
typedef: record deriving(eq, ord)
params: []
prefix: mylib_
cpp:
  typename: '::mylib::djinni::Record1'
  header: '"MyLib/djinni/Record1.hpp"'
  byValue: false # Controls 'const T&' vs. `T` (only used for records)
objc:
  typename: 'MLBRecord1'
  header: '"MLB/MLBRecord1.h"'
  param: 'MLBRecord1*' # Not required to be a pointer, e.g. NSTimeInterval (= double)
  boxed: 'MLBRecord1*' # Must be an object
objcpp:
  translator: '::mylib::djinni::objc::Record1'
  header: '"MyLib/djinni/objc/MLBRecord1+Private.h"'
java:
  typename: 'com.example.mylib.Record1'
jni:
  translator: '::mylib::djinni::jni::Record1'
  header: '"MyLib/djinni/jni/Record1.hpp"'
  typename: jobject # Must be a JNI object type, currently no support for primitives
  typeSignature: 'Lcom/example/mylib/Record1;'
---
name: mylib_interface1
typedef: interface +c
# ...
---
name: mylib_enum1
typedef: enum
# ...
```
It tries to preserve all the information necessary to integrate a type into another. All fields are mandatory. There is nothing stopping you from creating such a file by hand to have the translation follow different patterns, e.g. by specifying `std` types. I'm pretty sure this could be used to declare most of the builtin Djinni types (except containers due to deep copying in Obj-C and `optional` because of the very special template parameter). It is your responsibility to ensure all the header paths remain valid if used in your project. This currently also includes the Djinni support library as in JNI enums don't have any custom translator class.

The translator types must be provided manually if not generated by Djinni itself and have the following anatomy:

Objective-C <-> C++:
```cpp
namespace mylib { namespace djinni { namespace objc {
struct Record1
{
    using CppType = ::mylib::djinni::Record1;
    using ObjcType = MLBRecord1*;

    static CppType toCpp(ObjcType objc);
    static ObjcType fromCpp(const CppType& cpp);

    // Must have an embedded "Boxed" type
    // Option1: self-reference if no special boxing behavior is necessary
    using Boxed = Record1;
    // Option2: Special boxing treatment required
    // Allows for custom boxing behavior
    // e.g. NSTimeInterval <-> NSNumber* or CGPoint <-> NSValue*
    struct Boxed {
        using ObjcType = MLBBoxedRecord1*;
        static CppType toCpp(ObjcType objc);
        static ObjcType fromCpp(const CppType& cpp);
    };
};
} } }
```
JNI <-> C++:
```cpp
namespace mylib { namespace djinni { namespace jni {
struct Record1
{
    using CppType = ::mylib::djinni::Record1;
    using JniType = jobject;

    ~Record1();

    static CppType toCpp(JNIEnv* jniEnv, JniType objc);
    // IMPORTANT: Must return LocalRef<T> if not a primitive!
    static ::djinni::LocaLref<JniType> fromCpp(JNIEnv* jniEnv, const CppType& cpp);

    // Must have an embedded "Boxed" type, same procedure as above
    using Boxed = Record1;

private:
    Record1();
    friend ::djinni::JniClass<Record1>;
    // Whatever JNI cruft is necessary to load your class (not used by outside code)
    // This is very likely used inside your own toCpp/fromCpp methods
    const ::djinni::GlobalRef<jclass> clazz { ::djinni::jniFindClass("com/example/mylib/Record1") };
    const jmethodID jconstructor { ::djinni::jniGetMethodID(clazz.get(), "<init>", "(D)V") };
    const jfieldID field_x { ::djinni::jniGetFieldID(clazz.get(), "x", "D") };
};
} } }
```

Let me know what you think. There is still lots of room for improvement. The first version does not preserve enumerators or record fields so they cannot be used as constants. Template parameters are supported in theory but I have not tested it yet. I totally expect there to be edge cases where the extern integration is broken.

**Comment by skabbes** [comment]
I haven't taken a very deep look at this yet, but my first reaction is that this is amazing.  Solves lots of problems, especially when back-porting djinni onto an existing project, as I am :)

@mknejp You have a few of outstanding PR's with a lot of code, I wanted to make sure we get them in in the order you intended.  What do you suggest is the order of operations here?  You mentioned that this was based off refactoring for #54 but I didn't see any code there.

**Comment by mknejp** [comment]
I didn't create pull requests yet *because* it's such a big amount of code, so I just linked my branches. I could go ahead and do pull requests if you want, but was thinking about getting feedback first and make sure it doesn't break anything.

**Comment by pwais** [comment]
@mknejp hawt stuff! I haven't looked too deep either (beyond the new Primitives) but thanks for hacking on this!!!  Very excited to try this out in the context of the earlier buffer discussion.  

Quick question: Does this change require c++14 or will c++11 work?  (Perhaps there's some template feature issues in Marshal ... ? )

**Comment by mknejp** [comment]
Ah yes, I think Marshal is using C++14 deduced return types. Mainly because I'm lazy and my brain is in C++14 mode. Should be easy to change if it's a blocker.

**Comment by skabbes** [comment]
No need to do PRs just yet, just trying to understand all the dependencies of this.  I'll hopefully take a look tomorrow, but really excited about this in general.

**Comment by j4cbo** [comment]
I'm fine with moving to C++14 for Djinni - we already build everything in '14 mode here. @pwais would that be a blocker for you or is just an issue of changing compilation flags?

**Comment by pwais** [comment]
I only see only 6 lines in Marshal.hpp that use c++14-style auto return type deduction... couldn't those be easily converted to c++11-style (e.g. `static auto unbox(..)` to `static auto unbox(..) -> jlong`) ?

I can't get everything in the branch to compile, but it it seems at least the ObjC code is fine in c++11 mode.  (I changed `test-suite/objc/DjinniObjcTest.xcodeproj/project.pbxproj` to use c++11.  Looking briefly at that code I only found c++11-style auto-deduced return types, i.e. with a `-> returntype` specifier).

That said, I don't believe c++14 is a blocker for my use case.  My main concern would be linking with binaries compiled for c++11 but I don't believe there should be a problem.

**Comment by mknejp** [comment]
Actually, those `auto`s can all be replaced with `JniType` as that is exactly what the unboxing functions are supposed to return

**Comment by mknejp** [comment]
The trailing return types are there to avoid parsing ambiguities. In
```cpp
CppType ::my::djinni::Type::toCpp(...)
```
the compiler assumes the first `::` is indexing `CppType`, therefore the change to
```cpp
auto ::my::djinni::Type::toCpp(...) -> CppType
```
But this is C++11 only, which is required for many other reasons.

**Comment by j4cbo** [comment]
This is fixed by #95 :)
